vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+        b.nodes[cn].right = right_child;
+        b.nodes[left_child].parent = cn;
+        b.nodes[right_child].parent = cn;
+        b.nodes[left_child].depth = b.nodes[cn].depth + 1;
+        b.nodes[right_child].depth = b.nodes[cn].depth + 1;
+
+        if (b.nodes[left_child].depth > b.max_depth)
+            b.max_depth = b.nodes[left_child].depth;
+
+        //rout the examples stored in the node to the left and right
+        for(size_t ec_id = 0; ec_id < b.nodes[cn].examples_index.size(); ec_id++) //scan all examples stored in the cn
+        {
+            uint32_t ec_pos = b.nodes[cn].examples_index[ec_id];
+            MULTICLASS::label_t mc;
+            uint32_t save_multi_pred;
+            MULTILABEL::labels multilabels;
+            MULTILABEL::labels preds;
+            if (b.oas == false){
+                mc = b.examples[ec_pos]->l.multi;
+                save_multi_pred = b.examples[ec_pos]->pred.multiclass;
+            }
+            else{
+                multilabels = b.examples[ec_pos]->l.multilabels;
+                preds = b.examples[ec_pos]->pred.multilabels;
+            }
+
+            b.examples[ec_pos]->l.simple = {1.f, 1.f, 0.f};
+            base.predict(*b.examples[ec_pos], b.nodes[cn].base_router); //re-predict
+            float scalar = b.examples[ec_pos]->pred.scalar;
+            if (scalar < 0)
+            {
+                b.nodes[left_child].examples_index.push_back(ec_pos);
+                float leaf_pred = train_node(b, base, *b.examples[ec_pos], left_child);
+                descent(b.nodes[left_child], leaf_pred); //fake descent, only for update nl and nr                
+            }
+            else
+            {
+                b.nodes[right_child].examples_index.push_back(ec_pos);
+                float leaf_pred = train_node(b, base, *b.examples[ec_pos], right_child);
+                descent(b.nodes[right_child], leaf_pred); //fake descent. for update nr and nl
+            }
+
+            if(b.oas == false){
+                b.examples[ec_pos]->l.multi = mc;
+                b.examples[ec_pos]->pred.multiclass = save_multi_pred;
+            }
+            else{
+                b.examples[ec_pos]->pred.multilabels = preds;
+                b.examples[ec_pos]->l.multilabels = multilabels;
+            }
+        }
+        b.nodes[cn].examples_index.delete_v(); //empty the cn's example list
+        b.nodes[cn].nl = std::max(double(b.nodes[left_child].examples_index.size()), 0.001); //avoid to set nl to zero
+        b.nodes[cn].nr = std::max(double(b.nodes[right_child].examples_index.size()), 0.001); //avoid to set nr to zero
+
+        if (std::max(b.nodes[cn].nl, b.nodes[cn].nr) > b.max_ex_in_leaf)
+        {
+            b.max_ex_in_leaf = std::max(b.nodes[cn].nl, b.nodes[cn].nr);
+            //cout<<b.max_ex_in_leaf<<endl;
+        }
+    }
+
+
+    template<typename T> 
+    inline bool in_v_array(const v_array<T>& array, const T& item, uint32_t& pos){
```

Comment:
Is this equivalent to v_array_contains() https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/v_array.h#L245